wit/types.wit

```diff
@@ -0,0 +1,158 @@
+  fields-clone: func(fields: fields) -> fields
+
+  type headers = fields
+  type trailers = fields
+
+  // The following block defines the `body` type which corresponds to the HTTP
+  // standard Contents. With Preview3, all of these fields can be replaced by a
+  // single type definition:
+  //
+  //   type body = stream<u8, option<trailers>>
+  //
+  // In the interim, we need to use separate `input-stream` and `output-stream`
+  // resource types defined by `wasi:io/streams`. The `finish-` functions
+  // emulate the stream's result value and MUST be called exactly once after
+  // the final read/write from/to the stream before dropping the stream.
+  type incoming-body = input-stream
+  type outgoing-body = output-stream
+  finish-incoming-body: func(body: incoming-body) -> option<trailers>
+  finish-outgoing-body: func(body: outgoing-body, trailers: option<trailers>)
+
+  // The following block defines the `incoming-request` and `outgoing-request`
+  // resource types that correspond to HTTP standard Requests. Soon, when
+  // resource types are added, the `u32` type aliases can be replaced by
+  // proper `resource` type definitions containing all the functions as
+  // methods. Later, Preview2 will allow both types to be merged together into
+  // a single `request` type (that uses the single `body` type mentioned
+  // above). The `consume` and `write-body` methods may only be called once
+  // (and return failure thereafter).
+  type incoming-request = u32
+  type outgoing-request = u32
+  drop-incoming-request: func(request: incoming-request)
+  drop-outgoing-request: func(request: outgoing-request)
+  incoming-request-method: func(request: incoming-request) -> method
+  incoming-request-path: func(request: incoming-request) -> string
+  incoming-request-scheme: func(request: incoming-request) -> option<scheme>
+  incoming-request-authority: func(request: incoming-request) -> string
+  incoming-request-headers: func(request: incoming-request) -> headers
+  incoming-request-consume: func(request: incoming-request) -> result<incoming-body>
+  new-outgoing-request: func(
+    method: method,
+    path: string,
+    scheme: option<scheme>,
+    authority: string,
+    headers: headers
+  ) -> outgoing-request
+  outgoing-request-write-body: func(request: outgoing-request) -> result<outgoing-body>
+
+  // Additional optional parameters that can be set when making a request.
+  record request-options {
+    // The following timeouts are specific to the HTTP protocol and work
+    // independently of the overall timeouts passed to `io.poll.poll-oneoff`.
+
+    // The timeout for the initial connect.
+    connect-timeout-nanos: option<u64>,
+
+    // The timeout for receiving the first byte of the response body.
+    first-byte-timeout-ns: option<u64>,
+
+    // The timeout for receiving the next chunk of bytes in the response body
+    // stream.
+    between-bytes-timeout-ms: option<u32>
```

Comment:
Those feel very inconsistent with different suffixes, granularity and value types. What's the reason behind that?

Reply:
Oh gosh, that's embarrassing; it's like I forgot to `git add` before a push or something :P  I aligned them on `ms` [here](https://github.com/WebAssembly/wasi-http/pull/3/commits/297f6536c893f8c316966bc179046523de7280fb), but I think we can discuss further in a follow-up issue if we should change that.